File: app/src/main/java/md/fusionworks/lifehack/ui/LoadMoreAdapter.java
package md.fusionworks.lifehack.ui;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import java.util.List;

/**
 * Created by ungvas on 12/30/15.
 */
public abstract class LoadMoreAdapter<T> extends ItemListAdapter<T> {

  private final T ITEM_LOAD_MORE_VIEW = null;

  private OnLoadMoreItemsListener onLoadMoreItemsListener;
  private boolean isLoading;
  private int visibleThreshold = 5;
  private int lastVisibleItem, totalItemCount;

  public LoadMoreAdapter(RecyclerView recyclerView) {
    super();
    LinearLayoutManager linearLayoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
    recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
      @Override public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
        super.onScrolled(recyclerView, dx, dy);

        totalItemCount = linearLayoutManager.getItemCount();
        lastVisibleItem = linearLayoutManager.findLastVisibleItemPosition();

        if (!isLoading && totalItemCount <= (lastVisibleItem + visibleThreshold)) {
          if (onLoadMoreItemsListener != null) {
            recyclerView.post(() -> addLoadMoreViewItem());
            onLoadMoreItemsListener.onLoadMoreItems();
          }
          isLoading = true;
        }
      }
    });
  }

  public void setLoaded() {
    isLoading = false;
  }

  /**
   * Add special item (null) that indicate to show load more items view
   */
  private void addLoadMoreViewItem() {
    getItemList().add(ITEM_LOAD_MORE_VIEW);
    notifyItemInserted(getItemCount() - 1);
  }

  /**
   * Remove special item (null) that indicate to show load more items view
   */
  private void removeLoadMoreViewItem() {
    if (getItemList().size() > 0) {
      getItemList().remove(getItemList().size() - 1);
      notifyItemRemoved(getItemCount());
    }
  }

  @Override public void addItems(List<T> itemList) {
    removeLoadMoreViewItem();
    super.addItems(itemList);
    setLoaded();
  }

  public void setOnLoadMoreItemsListener(OnLoadMoreItemsListener onLoadMoreItemsListener) {
    this.onLoadMoreItemsListener = onLoadMoreItemsListener;
  }
}

File: app/src/main/java/md/fusionworks/lifehack/ui/base/view/LoadMoreAdapter.kt
package md.fusionworks.lifehack.ui.base.view

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView

/**
 * Created by ungvas on 12/30/15.
 */
abstract class LoadMoreAdapter<T>(recyclerView: RecyclerView) : ItemListAdapter<T>() {

  private val ITEM_LOAD_MORE_VIEW: T? = null

  private var onLoadMoreItemsListener: OnLoadMoreItemsListener? = null
  private var isLoading: Boolean = false
  private val visibleThreshold = 5
  private var lastVisibleItem: Int = 0
  private var totalItemCount: Int = 0

  init {
    val linearLayoutManager = recyclerView.layoutManager as LinearLayoutManager
    recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
      override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
        super.onScrolled(recyclerView, dx, dy)

        totalItemCount = linearLayoutManager.itemCount
        lastVisibleItem = linearLayoutManager.findLastVisibleItemPosition()

        if (!isLoading && totalItemCount <= lastVisibleItem + visibleThreshold) {
          if (onLoadMoreItemsListener != null) {
            recyclerView!!.post { addLoadMoreViewItem() }
            onLoadMoreItemsListener!!.onLoadMoreItems()
          }
          isLoading = true
        }
      }
    })
  }

  fun setLoaded() {
    isLoading = false
  }

  /**
   * Add special item (null) that indicate to show load more items view
   */
  private fun addLoadMoreViewItem() {
    items.add(ITEM_LOAD_MORE_VIEW)
    notifyItemInserted(itemCount - 1)
  }

  /**
   * Remove special item (null) that indicate to show load more items view
   */
  private fun removeLoadMoreViewItem() {
    if (items.size > 0) {
      items.removeAt(items.size - 1)
      notifyItemRemoved(itemCount)
    }
  }

  override fun addItems(itemList: List<T>) {
    removeLoadMoreViewItem()
    super.addItems(itemList)
    setLoaded()
  }

  fun setOnLoadMoreItemsListener(onLoadMoreItemsListener: OnLoadMoreItemsListener) {
    this.onLoadMoreItemsListener = onLoadMoreItemsListener
  }

  interface OnLoadMoreItemsListener {
    fun onLoadMoreItems()
  }
}
